src/DynamoCore/Core/TrustedLocatationsManager.cs

```diff
@@ -0,0 +1,153 @@
+﻿using Dynamo.Configuration;
+using DynamoUtilities;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace Dynamo.Core
+{
+    /// <summary>
+    /// Manages trusted locations for Dynamo
+    /// </summary>
+    public class TrustedLocatationsManager
+    {
+        /// <summary>
+        /// Use Lazy&lt;TrustedLocatationsManager&gt; to make sure the Singleton class is only initialized once
+        /// </summary>
+        internal static readonly Lazy<TrustedLocatationsManager>
+            lazy =
+            new Lazy<TrustedLocatationsManager>
+            (() => new TrustedLocatationsManager());
+
+        private readonly PreferenceSettings settings;
+
+        /// <summary>
+        /// Dynamo's trusted locations.
+        /// Dynamo will load .dyn files from these locations without asking the user for consent.
+        /// </summary>
+        private List<string> trustedLocations;
+
+        /// <summary>
+        /// An enumerable that represents Dynamo's trusted locations
+        /// </summary>
+        public IEnumerable<string> TrustedLocations => trustedLocations;
+
+        internal void Initialize(PreferenceSettings settings)
+        {
+            settings = settings ?? throw new ArgumentNullException(nameof(settings));
+            trustedLocations = settings.TrustedLocations?.ToList() ?? new List<string>();
+        }
+
+        #region Public members
+        /// <summary>
+        /// The actual instance stored in the Singleton class
+        /// </summary>
+        public static TrustedLocatationsManager Instance
+        {
+            get { return lazy.Value; }
+        }
+
+        /// <summary>
+        /// Add a path to the Dynamo's trusted locations
+        /// </summary>
+        /// <param name="path">The path to be added as a trusted location</param>
+        /// <returns></returns>
+        public bool AddTrustedLocation(string path)
+        {
+            try
+            {
+                string location = ValidateTrustedLocation(path);
+                var existing = TrustedLocations.FirstOrDefault(x => x.Equals(location));
+                if (string.IsNullOrEmpty(existing))
+                {
+                    trustedLocations.Add(location);
+                    settings.TrustedLocations = trustedLocations;
+                    return true;
```

Comment:
Adding a trusted location will not allow duplicates...but it will allow subdirectories of existing trusted locations to be added .
Ex. of adding a new location:
Existing location : C:/programFiles/Autodesk
Incoming location: C:/programFiles/Autodesk/Revit
--------------------------------------------------------------
Locations : C:/programFiles/Autodesk, C:/programFiles/Autodesk/Revit

Reply:
made Add/Remove methods as internal